### PR TITLE
[TOOLS-2694] Touch Time Add Timeline Filter

### PR DESCRIPTION
**Pull request by marcokuchla**

### CODE REVIEW

File: kpi-parent/kpi-testutils/src/test/java/objective/taskboard/followup/kpi/services/KpiDataServiceTest.java [hunk 1]
```diff
@@ -544,10 +658,9 @@ public void behave(KpiEnvironment environment) {
         protected KpiDataService getService(KpiEnvironment environment) {
             IssueKpiService issueKpiService = environment.services().issueKpi().getService();
             FollowUpSnapshotService followupSnapshotService = environment.services().followupSnapshot().getService();
-            ProjectService projectService = environment.services().projects().getService();
             KpiFilterService filterService = environment.services().filters().getService();
             
-            return new KpiDataService(issueKpiService, followupSnapshotService, projectService, filterService);
+            return new KpiDataService(issueKpiService, followupSnapshotService, filterService);
```
Comment (herbert-venancio) [inline kpi-parent/kpi-testutils/src/test/java/objective/taskboard/followup/kpi/services/KpiDataServiceTest.java]:
```java
            return environment.services().kpiDataService().getImplementation();
```
?

File: kpi-parent/kpi-testutils/src/test/java/objective/taskboard/followup/kpi/services/KpiDataServiceTest.java [hunk 2]
```diff
@@ -506,6 +516,97 @@ public void getIssuesFromCurrentState_allIssues() {
...
+                        .status("Open").date("2020-01-01")
+                        .status("To Do").noDate()
+                        .status("Doing").noDate()
+                        .status("Done").noDate()
+                    .eoT()
+                .endOfSubtask()
+            .eoI()
+            .when()
+                .appliesBehavior(
+                        getIssuesGivenFilters(
```
Comment (herbert-venancio) [inline kpi-parent/kpi-testutils/src/test/java/objective/taskboard/followup/kpi/services/KpiDataServiceTest.java]:
``` java
                .appliesBehavior(
                        getIssues()
                                .forProject("PROJ")
                                .filterLevel(SUBTASKS)
                                .filterProgressingDateRangeFilter(UTC_ZONE_ID, getTimeline("2020-01-06", "2020-01-10")))
```

Comment (herbert-venancio) [reply kpi-parent/kpi-testutils/src/test/java/objective/taskboard/followup/kpi/services/KpiDataServiceTest.java]: \+ this
```java
    private static class ServeIssuesFromCurrentStateBuilder implements DSLSimpleBehaviorWithAsserter<IssuesAsserter> {

        private String projectKey;
        private final List<KpiItemFilter> filters = new ArrayList<>();
        private ServeIssuesGivenFilters delegate;

        public ServeIssuesFromCurrentStateBuilder forProject(String projectKey) {
            this.projectKey = projectKey;
            return this;
        }

        public ServeIssuesFromCurrentStateBuilder filterLevel(KpiLevel level) {
            filters.add(new KpiLevelFilter(level));
            return this;
        }

        public ServeIssuesFromCurrentStateBuilder filterProgressingDateRangeFilter(ZoneId timezone, DashboardTimeline timeline) {
            ProjectTimelineRange range = new ProjectTimelineRangeByDashboardTimeline(timeline);
            filters.add(new KpiProgressingDateRangeFilter(timezone, range));
            return this;
        }

        @Override
        public void behave(KpiEnvironment environment) {
            delegate = build();
            delegate.behave(environment);
        }

        @Override
        public IssuesAsserter then() {
            return delegate.then();
        }

        private ServeIssuesGivenFilters build() {
            return new ServeIssuesGivenFilters(projectKey, filters.stream().toArray(KpiItemFilter[]::new));
        }
    }
```

File: kpi-parent/kpi-touch-time/src/test/java/objective/taskboard/followup/kpi/touchtime/helpers/GenerateTTByWeekDataSetStrategyBehavior.java [hunk 3]
```diff
@@ -11,7 +12,10 @@
 import objective.taskboard.followup.kpi.touchtime.TouchTimeByWeekKpiDataPoint;
 import objective.taskboard.followup.kpi.touchtime.TouchTimeByWeekKpiStrategy;
 import objective.taskboard.followup.kpi.touchtime.TouchTimeByWeekKpiStrategyFactory;
+import objective.taskboard.jira.ProjectService;
 import objective.taskboard.jira.properties.JiraProperties;
+import objective.taskboard.timeline.DashboardTimelineProvider;
+import objective.taskboard.utils.Clock;
 
 public class GenerateTTByWeekDataSetStrategyBehavior extends GenerateTTDataSetStrategyBehavior<TTByWeekKpiDataSetAsserter> {
```
[the remark this answers is not in the document]
Comment (marcokuchla) [reply kpi-parent/kpi-touch-time/src/test/java/objective/taskboard/followup/kpi/touchtime/helpers/GenerateTTByWeekDataSetStrategyBehavior.java]: Done.

File: kpi-parent/kpi-testutils/src/main/java/objective/taskboard/followup/kpi/services/MockedServices.java [hunk 4]
```diff
@@ -400,6 +410,44 @@ private void mockStatusFilter() {
...
+        public DashboardConfigurationService getService() {
+            if (service == null) {
+                prepareMock();
+            }
+            return service;
+        }
+        private void prepareMock() {
+            service = Mockito.mock(DashboardConfigurationService.class);
+            projects.getProjectsKeys().forEach(
+                    pk -> Mockito.when(service.retrieveConfiguration(pk)).thenReturn(getConfiguration(pk)));
```
[the remark this answers is not in the document]
Comment (marcokuchla) [reply kpi-parent/kpi-testutils/src/main/java/objective/taskboard/followup/kpi/services/MockedServices.java]: Done.

File: kpi-parent/kpi-testutils/src/main/java/objective/taskboard/followup/kpi/services/ProjectServiceMocker.java [hunk 5]
```diff
@@ -43,9 +43,21 @@ public ProjectFilterConfiguration getProject(String projectKey) {
...
+        projects.values().forEach(p -> p.mockProject(projectService));
+    }
+
+    private void configureServiceToDefaultBehavior() {
+        configureExceptionToNotConfiguredProjects();
+        configureReturnOptionalEmptyUnconfiguredProjects();
+    }
 
+    private void configureReturnOptionalEmptyUnconfiguredProjects() {
+        Mockito.when(projectService.getTaskboardProject(Mockito.anyString())).thenReturn(Optional.empty());
```
[the remark this answers is not in the document]
Comment (marcokuchla) [reply kpi-parent/kpi-testutils/src/main/java/objective/taskboard/followup/kpi/services/ProjectServiceMocker.java]: Done.

File: kpi-parent/kpi-testutils/src/test/java/objective/taskboard/followup/kpi/services/KpiDataServiceTest.java [hunk 6]
```diff
@@ -506,6 +516,97 @@ public void getIssuesFromCurrentState_allIssues() {
...
+            .withTimezone(UTC_ZONE_ID)
+            .services()
+                .projects()
+                    .withKey("PROJ")
+                        .startAt("2020-01-04")
+                        .deliveredAt("2020-01-10")
+                    .eoP()
+                .eoPs()
+            .eoS()
+            .givenFeature("I-1")
```
[the remark this answers is not in the document]
Comment (marcokuchla) [reply kpi-parent/kpi-testutils/src/test/java/objective/taskboard/followup/kpi/services/KpiDataServiceTest.java]: Done.

File: kpi-parent/kpi-touch-time/src/main/java/objective/taskboard/followup/kpi/touchtime/TouchTimeByIssueKpiStrategyFactory.java [hunk 7]
```diff
@@ -26,23 +32,32 @@
     private KpiDataService kpiDataService;
     private JiraProperties jiraProperties;
     private MetadataService metadataService;
```
[the remark this answers is not in the document]
Comment (marcokuchla) [reply kpi-parent/kpi-touch-time/src/main/java/objective/taskboard/followup/kpi/touchtime/TouchTimeByIssueKpiStrategyFactory.java]: Fixed.

File: kpi-parent/kpi-touch-time/src/test/java/objective/taskboard/followup/kpi/touchtime/TouchTimeByWeekKpiStrategyIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        dsl.environment()
+            .types()
+                .addFeatures("Task")
+                .addSubtasks("Backend Development", "Alpha Bug", "Tech Analysis", "Subtask Review")
+            .eoT()
+            .statuses()
+                .withNotProgressingStatuses("Open", "To Plan", "To Do", "To Review", "Done")
+                .withProgressingStatuses("Planning", "Doing", "Reviewing")
+            .eoS()
+            .withTimezone(ZONE_ID);
```
[the remark this answers is not in the document]
Comment (marcokuchla) [reply kpi-parent/kpi-touch-time/src/test/java/objective/taskboard/followup/kpi/touchtime/TouchTimeByWeekKpiStrategyIntegrationTest.java]: Fixed.

File: kpi-parent/kpi-touch-time/src/test/java/objective/taskboard/followup/kpi/touchtime/helpers/GenerateTTByWeekDataSetStrategyITBehavior.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+import objective.taskboard.followup.kpi.services.KpiEnvironment;
+import objective.taskboard.followup.kpi.touchtime.TouchTimeByWeekKpiDataPoint;
+import objective.taskboard.followup.kpi.touchtime.TouchTimeByWeekKpiStrategy;
+import objective.taskboard.followup.kpi.touchtime.TouchTimeByWeekKpiStrategyFactory;
+import objective.taskboard.jira.ProjectService;
+import objective.taskboard.jira.properties.JiraProperties;
+import objective.taskboard.timeline.DashboardTimelineProvider;
+import objective.taskboard.utils.Clock;
+
+public class GenerateTTByWeekDataSetStrategyITBehavior extends GenerateTTDataSetStrategyBehavior<TTByWeekKpiDataSetAsserter> {
```
[the remark this answers is not in the document]
Comment (marcokuchla) [reply kpi-parent/kpi-touch-time/src/test/java/objective/taskboard/followup/kpi/touchtime/helpers/GenerateTTByWeekDataSetStrategyITBehavior.java]: Removed ITBehavior.

File: kpi-parent/kpi-testutils/src/test/java/objective/taskboard/followup/kpi/services/KpiDataServiceTest.java (hunk 2 above)
Comment (marcokuchla) [reply kpi-parent/kpi-testutils/src/test/java/objective/taskboard/followup/kpi/services/KpiDataServiceTest.java]: Done.

File: kpi-parent/kpi-testutils/src/test/java/objective/taskboard/followup/kpi/services/KpiDataServiceTest.java (hunk 1 above)
Comment (marcokuchla) [reply kpi-parent/kpi-testutils/src/test/java/objective/taskboard/followup/kpi/services/KpiDataServiceTest.java]: Fixed.

File: kpi-parent/kpi-touch-time/src/main/java/objective/taskboard/followup/kpi/touchtime/TouchTimeByWeekKpiStrategyFactory.java [hunk 10]
```diff
@@ -24,36 +29,41 @@
     private KpiTouchTimeProperties touchTimeProperties;
     private KpiDataService kpiDataService;
     private JiraProperties jiraProperties;
```
Comment (herbert-venancio) [inline kpi-parent/kpi-touch-time/src/main/java/objective/taskboard/followup/kpi/touchtime/TouchTimeByWeekKpiStrategyFactory.java]: why not final on all of them to be consistent?

File: kpi-parent/kpi-touch-time/src/main/java/objective/taskboard/followup/kpi/touchtime/TouchTimeByWeekKpiStrategyFactory.java [hunk 11]
```diff
@@ -24,36 +29,41 @@
...
+        this.dashboardTimelineProvider = dashboardTimelineProvider;
     }
 
     @Override
     public TouchTimeByWeekKpiStrategy getStrategy(KpiLevel level, ProjectFilterConfiguration projectConfiguration, ZoneId timezone) {
-        List<IssueKpi> issues = kpiDataService.getIssuesFromCurrentProjectRange(
-                projectConfiguration.getProjectKey(), timezone, level);
-        Range<LocalDate> projectRange = getRangeOrCry(projectConfiguration);
-
+        String projectKey = projectConfiguration.getProjectKey();
```
Comment (herbert-venancio) [inline kpi-parent/kpi-touch-time/src/main/java/objective/taskboard/followup/kpi/touchtime/TouchTimeByWeekKpiStrategyFactory.java]: This code is repeated:
```java
        String projectKey = projectConfiguration.getProjectKey();
        DashboardTimeline timeline = dashboardTimelineProvider.getDashboardTimelineForProject(projectKey, timezone);
        List<IssueKpi> issues = kpiDataService.getIssuesFromCurrentGivenFilters(
                projectKey,
                timezone,
                new KpiLevelFilter(level),
                new KpiProgressingDateRangeFilter(timezone, range));
```
Move it to a base class that both `TouchTimeByIssueKpiStrategyFactory` and `TouchTimeByWeekKpiStrategyFactory` extends
Also move `getProgressingStatuses` to base class

File: kpis/src/main/java/objective/taskboard/followup/kpi/services/KpiDataService.java [hunk 12]
```diff
@@ -60,17 +51,13 @@ public KpiDataService(
...
-    public List<IssueKpi> getIssuesFromCurrentProjectRange(String projectKey, ZoneId timezone, KpiLevel kpiLevel){
-        List<IssueKpi> allIssues = getIssuesFromCurrentStateWithDefaultFilters(projectKey, timezone, kpiLevel);
-        ProjectFilterConfiguration project =  projectService.getTaskboardProjectOrCry(projectKey);
-        ProjectTimelineRange range = new ProjectRangeByConfiguration(project);
-        
-        return allIssues.stream().filter(new WithinRangeFilter(timezone, range)).collect(Collectors.toList());
+        return concatFilters(statusFilter, levelFilter);
     }
 
+    private KpiItemFilter concatFilters(KpiItemFilter ...filters) {
```
[the remark this answers is not in the document]
Comment (marcokuchla) [reply kpis/src/main/java/objective/taskboard/followup/kpi/services/KpiDataService.java]: Done.

File: kpis/src/main/java/objective/taskboard/followup/kpi/services/KpiDataService.java [hunk 13]
```diff
@@ -83,4 +70,10 @@ public FollowUpSnapshot getSnapshot(String projectKey, ZoneId timezone, Optional
         return issueKpiService.getIssues(analyticTransitionDs);
     }
 
+    public List<IssueKpi> getIssuesFromCurrentGivenFilters(String projectKey, ZoneId timezone, KpiItemFilter ...filters) {
```
[the remark this answers is not in the document]
Comment (marcokuchla) [reply kpis/src/main/java/objective/taskboard/followup/kpi/services/KpiDataService.java]: Done.

File: kpi-parent/kpi-touch-time/src/main/java/objective/taskboard/followup/kpi/touchtime/TouchTimeByWeekKpiStrategyFactory.java (hunk 10 above)
Comment (marcokuchla) [reply kpi-parent/kpi-touch-time/src/main/java/objective/taskboard/followup/kpi/touchtime/TouchTimeByWeekKpiStrategyFactory.java]: Fixed.

**Pull request merged**